src/de/received_packet.rs

```diff
@@ -44,14 +71,28 @@ impl<'a> ReceivedPacket<'a> {
             }
         }
 
+        Ok(packet.packet)
+    }
+
+    /// Get a received packet from a split buffer
+    ///
+    /// # Note
+    /// Packets deserialized in this manner may not have valid binary slices. This occurs because
+    /// the binary slice may span the discontinuity, and thus it's impossible to zero-copy it.
+    /// Instead, the resulting binary slice will only contain the maximum amount of continuous
+    /// data.
```

Comment:
Being worried about how one would then go about actually getting the complete valid packet, it took me a while to figure out that this path is only used in republication, where we don't really need to deserialize anything other than the header for probing.
Wouldn't `header_only()` (and then returning after the header, or even only returning the header) instead of `truncate_on_discontinuity()` be more adequate?
And what happens if the header is split across the buffers? Does it still deserialize that correctly (since no borrowed data)?

Reply:
There's a few things to note:
1. When deserializing via the discontinuous buffer, the complete `packet` deserialized is _not_ `pub` to the crate, so it's not possible to unintentionally use invalid contents. The API only exposes `id()` and `len()` functions for these packets.
2. `header_only()` would be ideal, but unfortunately we need to get the `Pub` packet ID, which _may or may not_ be present for packets. So we have to do a complete fixed header deserialization, check the packet type, and then deserialize the `Pub` packet. We can go this route, but it's a lot of duplicated code
3. It's not relevant if the header is split across the slices because the header information contains only primitive types. That being said, I'm now seeing an issue where even u16s have to be within the continuous memory, will fix

Reply:
Ok. That kills the first alternative. What about the other (`Deque<(offset, MqttHeader), 10>` as an index into the `RingBuffer`)? It would only be slightly more code for managing that and a bit more stack but would remove the repeated deserialization and allow `take(n)` again in the deseroializer. It wouldn't duplicate any deser code.

Reply:
I got it even simpler with a single `Deque<offset, 10>` - no need for any deserialization until we're about to pop the packet.

Reply:
Nice. But there is still the one unnecessary deserialization in `pop_publish`. This is a packet that we serialized when it was added to the ring buffer, i.e. we know its packet length and packet id. Wouldn't removing that deserialization step (in favor of maintaining a `Deque<(usize, MqttHeader)>`) remove all the "deserialize from split buffer" gymnastics in `deserializer.rs`?